bpf/lib/ipv4.h

```diff
@@ -44,8 +68,71 @@ static __always_inline bool ipv4_is_fragment(struct iphdr *ip4)
 	// +----+----+----+----------------------------------+
 	//
 	// If "More fragments" or the offset is nonzero, then this is an IP
-	// fragment. The evil bit must be set to 0 (RFC791, RFC3514).
-	return ip4->frag_off & bpf_htons(0xBFFF);
+	// fragment (RFC791).
+	return ip4->frag_off & bpf_htons(0x3FFF);
+}
+
+#if defined IPV4_FRAGMENTS
+static __always_inline bool ipv4_is_not_first_fragment(const struct iphdr *ip4)
+{
+	/* Ignore "More fragments" bit to catch all fragments but the first */
+	return ip4->frag_off & bpf_htons(0x1FFF);
+}
+
+static __always_inline int
+ipv4_frag_get_l4ports(const struct ipv4_frag_id *frag_id,
+		      struct ipv4_frag_l4ports *ports)
+{
+	struct ipv4_frag_l4ports *tmp;
+
+	tmp = map_lookup_elem(&IPV4_FRAG_DATAGRAMS_MAP, frag_id);
+	if (!tmp)
+		return DROP_FRAG_NOSUPPORT;
+
+	/* Do not make ports a pointer to map data, copy from map */
+	__builtin_memcpy(ports, tmp, sizeof(*ports));
+	return 0;
+}
+
+static __always_inline int
+ipv4_frag_register_datagram(struct __ctx_buff *ctx, int l4_off,
+			    const struct ipv4_frag_id *frag_id,
+			    struct ipv4_frag_l4ports *ports)
+{
+	int ret;
+
+	ret = ctx_load_bytes(ctx, l4_off, ports, 4);
```

Comment:
Same here for direct packet copy. Plus, it looks like it would remove the need to pass `l4_off` as an argument to a few functions.

Reply:
Right, makes sense. I'll switch to direct access.